### PR TITLE
[SPARK-6979][Streaming] Replace JobScheduler.eventActor and JobGenerator.eventActor with EventLoop

### PR DESCRIPTION
**Pull request by zsxwing**
Title says it all.

cc @rxin @tdas

### CODE REVIEW

**Comment by srowen** [comment]
If you have a moment, this is more for my edification, is the benefit simply that there is no need to participate in the full-blown actor system, and that this is simpler to reason about as a local event loop? are there other benefits or drawabacks?

**Comment by zsxwing** [comment]
@srowen this is a part of [SPARK-5293](https://issues.apache.org/jira/browse/SPARK-5293). The future plan is make Spark be able to be built without Akka at all eventually.

**Comment by zsxwing** [comment]
retest this please

**Comment by zsxwing** [comment]
I just replaced `jobScheduler.reportError("Error in job generator", e)` with `logError("Error in job generator", e)` and `NotSerializableException` disappeared. Not sure why. Will dig deeply tomorrow.

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobGenerator.scala [hunk 1]
```diff
@@ -70,22 +68,23 @@ class JobGenerator(jobScheduler: JobScheduler) extends Logging {
...
-        case event: JobGeneratorEvent =>  processEvent(event)
-      }
-    }), "JobGenerator")
+    eventLoop = new EventLoop[JobGeneratorEvent]("JobGenerator") {
+      override protected def onReceive(event: JobGeneratorEvent): Unit = processEvent(event)
+
+      override protected def onError(e: Throwable): Unit = logError("Error in job generator", e)
+    }
+    eventLoop.start()
     if (ssc.isCheckpointPresent) {
```
Comment (rxin) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobGenerator.scala]: add a blank line here

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobScheduler.scala [hunk 2]
```diff
@@ -46,21 +48,20 @@ class JobScheduler(val ssc: StreamingContext) extends Logging {
...
-        case event: JobSchedulerEvent => processEvent(event)
-      }
-    }), "JobScheduler")
+    eventLoop = new EventLoop[JobSchedulerEvent]("JobScheduler") {
+      override protected def onReceive(event: JobSchedulerEvent): Unit = processEvent(event)
 
+      override protected def onError(e: Throwable): Unit = reportError("Error in job scheduler", e)
+    }
+    eventLoop.start()
     listenerBus.start(ssc.sparkContext)
```
Comment (rxin) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobScheduler.scala]: blank line here

**Comment by rxin** [comment]
Thanks - please ping me when this is ready.

**Comment by zsxwing** [comment]
Finally, figured out why `logError("Error in job generator", e)` won't fail the test cases. 

NotSerializableException also happens in the old codes, but it just forces the Actor restart because it's not caught by any code. Then I use EventLoop to replace Actor, so `NotSerializableException` will be caught and sent to the `onError` method. If I use `logError("Error in job generator", e)`, `NotSerializableException` is just logged and doesn't fail the tests. But if I use `jobScheduler.reportError("Error in job generator", e)`, the error will be propagated to the mail thread of the tests. That's why the tests fail.

`StreamingKMeans` must be `Serializable` since it will be used in the closure. 

So it's a bug that was hidden by Actor error mechanism. And now `jobScheduler.reportError` exposes the bug.

**Comment by zsxwing** [comment]
Just added `Serializable` to `StreamingKMeans`. @rxin if you think I need to open a JIRA for the `NotSerializableException` bug in `StreamingKMeans`, please let me know.

**Comment by zsxwing** [comment]
I think I should create a new PR for the NotSerializableException bug in StreamingKMeans so that we can merge it to branch 1.2 and 1.3.

**Comment by zsxwing** [comment]
Will update this PR once #5582 is merged.

**Comment by rxin** [comment]
LGTM. I'm going to merge it in master since it doesn't conflict.

**Pull request closed without merging**